staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/object_count_tracker.go

```diff
@@ -0,0 +1,133 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+// StorageObjectCountTracker is an interface that is used to keep track of
+// of the total number of objects for each resource.
+// {group}.{resource} is used as the key name to update and retrieve
+// the total number of objects for a given resource.
+type StorageObjectCountTracker interface {
+	// OnCount is invoked to update the current number of total
+	// objects for the given resource
+	OnCount(string, int64)
+
+	// Get returns the total number of objects for the given resource.
+	// If the given resource is not being tracked Get will return zero.
+	// For now, we do not differentiate between zero object count and
+	// a given resoure not being present.
+	Get(string) int64
+}
+
+// NewStorageObjectCountTracker returns an instance of
+// StorageObjectCountTracker interface that can be used to
+// keep track of the total number of objects for each resource.
+func NewStorageObjectCountTracker(stopCh <-chan struct{}) StorageObjectCountTracker {
+	const (
+		// type deletion (it applies mostly to CRD) is not a very frequent operation
+		// so we can afford to prune the cache at a large interval.
+		pruneInterval = 12 * time.Hour
+	)
+	tracker := &objectCountTracker{
+		clock:  &clock.RealClock{},
+		counts: map[string]*timestampedCount{},
+	}
+	go func() {
+		wait.PollUntil(
+			pruneInterval,
+			func() (bool, error) {
+				tracker.prune(pruneInterval)
+				return false, nil
+			}, stopCh)
+		klog.InfoS("StorageObjectCountTracker pruner is exiting")
+	}()
+
+	return tracker
+}
+
+// timestampedCount stores the count of a given resource with a last updated
+// timestamp so we can prune it after it goes stale for certain threshold.
+type timestampedCount struct {
+	count       int64
+	lastUpdated time.Time
+}
+
+// objectCountTracker implements StorageObjectCountTracker with
+// reader/writer mutual exclusion lock.
+type objectCountTracker struct {
+	clock clock.PassiveClock
+
+	lock   sync.RWMutex
+	counts map[string]*timestampedCount
+}
+
+func (t *objectCountTracker) OnCount(groupResource string, count int64) {
+	if count == -1 {
+		// a value of -1 indicates that the 'Count' call failed to contact
+		// the storage layer, in most cases this error can be transient.
+		// we will continue to work with the count that is in the cache
```

Comment:
I'm not sure this is the right approach - if this isn't transient for some reason, the we may be operating on very stale data...

Thinking more about this - maybe this particular codepath is fine. But I would like to avoid returning stale results.
So maybe we can modify the get path to return "unknown" or sth like in case the data is too old.

Reply:
> But I would like to avoid returning stale results.
So maybe we can modify the get path to return "unknown" or sth like in case the data is too old.

with the pruning period set to `1h` - if this is not a transient error then after 1h the count will get stale and be removed from the cache, and then Get will return `0, false`. I think we are good, in the worst case we will use stale data for an hour and then will continue to return 'count not known` until the issue gets fixed.